clang/lib/Driver/ToolChains/SYCL.cpp

```diff
@@ -143,9 +148,172 @@ void SYCL::Linker::ConstructJob(Compilation &C, const JobAction &JA,
+      // passed in -Xsycl-linker.
+      continue;
+
+    // Add the argument from -Xsycl-linker.
+    StringRef ArgString;
+    if (XSYCLLinkerNoTriple) {
+      // With multiple -fsycl-targets, a triple is required so we know where
+      // the options should go.
+      if (Args.getAllArgValues(options::OPT_fsycl_targets_EQ).size() != 1) {
+        C.getDriver().Diag(diag::err_drv_Xsycl_linker_missing_triple);
+        continue;
+      }
+      // No triple, so just add the argument.
+      ArgString = A->getValue();
+      //A->render(A->getValue(), CmdArgs);
+    } else
+      // Triple found, add the next argument in line.
+      ArgString = A->getValue(1);
+    // Do a simple parse of the args to pass back
+    SmallVector<StringRef, 16> TargetArgs;
+    ArgString.split(TargetArgs, ' ', -1, false);
+    for (const auto &TA : TargetArgs)
+      CmdArgs.push_back(Args.MakeArgString(TA));
+    Args.ClaimAllArgs(options::OPT_Xsycl_linker_EQ);
+    Args.ClaimAllArgs(options::OPT_Xsycl_linker);
+  }
+}
+
+void SYCL::fpga::BackendCompiler::ConstructJob(Compilation &C,
+                                         const JobAction &JA,
+                                         const InputInfo &Output,
+                                         const InputInfoList &Inputs,
+                                         const ArgList &Args,
+                                         const char *LinkingOutput) const {
+  assert((getToolChain().getTriple().getArch() == llvm::Triple::spir ||
+          getToolChain().getTriple().getArch() == llvm::Triple::spir64) &&
+         "Unsupported target");
+  ArgStringList CmdArgs{"-o",  Output.getFilename()};
+  for (const auto &II : Inputs) {
+    CmdArgs.push_back(II.getFilename());
+  }
+  CmdArgs.push_back("-sycl");
+  TranslateSYCLTargetArgs(C, Args, getToolChain(), CmdArgs);
+
+  SmallString<128> ExecPath(getToolChain().GetProgramPath("aoc"));
+  const char *Exec = C.getArgs().MakeArgString(ExecPath);
+  C.addCommand(llvm::make_unique<Command>(JA, *this, Exec, CmdArgs, None));
+}
+
+void SYCL::x86_64::BackendCompiler::ConstructJob(Compilation &C,
+                                         const JobAction &JA,
+                                         const InputInfo &Output,
+                                         const InputInfoList &Inputs,
+                                         const ArgList &Args,
+                                         const char *LinkingOutput) const {
+  assert((getToolChain().getTriple().getArch() == llvm::Triple::spir ||
+          getToolChain().getTriple().getArch() == llvm::Triple::spir64) &&
+         "Unsupported target");
+  ArgStringList CmdArgs;
+  CmdArgs.push_back(Args.MakeArgString(Twine("-ir=") + Output.getFilename()));
+  CmdArgs.push_back("-cmd=link");
+  CmdArgs.push_back("-device=cpu");
+  for (const auto &II : Inputs) {
+    CmdArgs.push_back(Args.MakeArgString(Twine("-binary=") + II.getFilename()));
```

Comment:
What are the `Inputs` here? LLVM IR Bitcode with spir64 triple?

Reply:
> What are the `Inputs` here? LLVM IR Bitcode with spir64 triple?

Yes, that is the expected input for ioc.  Depending on expectations of the tool, this can either be LLVM-IR or SPIR-V.

Reply:
This may work for now, but LLVM IR is not a stable interface and older/newer CPU runtime may not consume LLVM IR generated by SYCL clang.

Reply:
Be optimistic... :-)